### PR TITLE
feat(uci): add uci storage module

### PR DESCRIPTION
**Pull request by benmix**

### CODE REVIEW

**Comment by benmix** [comment]
Work in process, need to add uci config parser. That work is in doing.

File: Cargo.toml [hunk 1]
```diff
@@ -12,4 +12,4 @@ http = []
 [dependencies]
 smol = "1.2.5"
 json = "0.12.4"
-
+tempfile = "3"
```
Comment (jerrykingxyz) [inline Cargo.toml]: why we need tempfile as a dependency? the std::fs seems to be enough

File: src/storage/uci/mod.rs [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+use tempfile::NamedTempFile;
+
+use uci_config::UciConfig;
+use uci_option::UciOption;
+use uci_section::UciSection;
+
+use self::uci_option::UciOptionType;
+
+pub struct UciTree {
+    dir: Box<Path>,
```
Comment (jerrykingxyz) [inline src/storage/uci/mod.rs]: the dir maybe not need to be wrapped in Box

File: src/storage/uci/mod.rs [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+impl UciTree {
+    pub fn new(root: String) -> Self {
+        Self {
+            dir: Box::from(Path::new(&root)),
+            configs: HashMap::new(),
+            lock: Mutex::new(false),
+        }
+    }
+
+    fn _load_config(&self, name: &str) -> Result<(), Error> {
```
Comment (jerrykingxyz) [inline src/storage/uci/mod.rs]: crate::error::Result ?

File: src/storage/uci/mod.rs [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+        };
+        match self.load_config(config_name) {
+            Ok(_) => {
+                if let Some(cfg) = self.configs.get(config_name) {
+                    Ok(cfg)
+                } else {
+                    Err(Error::new(format!("load config {} fail", config_name)))
+                }
+            }
+            Err(err) => Err(err),
```
Comment (jerrykingxyz) [inline src/storage/uci/mod.rs]: use question operator?

File: src/storage/uci/mod.rs [hunk 5]
```diff
@@ -0,0 +1,418 @@
...
+                Ok(Some(section)) => Ok(section.get(option_name)),
+                Ok(None) => Ok(None),
+                Err(err) => Err(err),
+            },
+            None => Ok(None),
+        }
+    }
+}
+
+pub trait Uci {
```
Comment (jerrykingxyz) [inline src/storage/uci/mod.rs]: this trait only use in UciTree?

Comment (benmix) [reply src/storage/uci/mod.rs]: Yes, I'll change

File: src/storage/uci/mod.rs (hunk 4 above)
Comment (benmix) [reply src/storage/uci/mod.rs]: Yes, I'll change

File: src/storage/uci/mod.rs (hunk 3 above)
Comment (benmix) [reply src/storage/uci/mod.rs]: Get, I'll change it to crate::utils::Result

File: src/storage/uci/mod.rs (hunk 2 above)
Comment (benmix) [reply src/storage/uci/mod.rs]: because path does not implement 'Sized' trait

File: Cargo.toml (hunk 1 above)
Comment (benmix) [reply Cargo.toml]: I try it

**Comment by benmix** [comment]
Finish uci module

**Pull request closed without merging**